### ISSUE TITLE
"CG Terminated in 1000 iterations with average residual norm ____ , which is larger than the tolerance of 1 specified by gpytorch.settings.cg_tolerance)" Warning- What does this mean?

### ISSUE THREAD

**Issue by christae** [cornellius-gp/gpytorch#1129]
I am currently attempting to fit a multi-task GPR (6 outputs, 16 inputs) with a multi-task RBF kernel from GPyTorch. When I use pytorch optimizer (adam, LBFGS), I get a bunch of warnings of the form from the title: "CG Terminated in 1000 iterations with average residual norm ____ , which is larger than the tolerance of 1 specified by gpytorch.settings.cg_tolerance)". 

Interestingly, these warnings do not show up for maybe the first couple of iterations of the optimization process, and the training loss predictably goes down as we increase the iterations. At a certain point, however, I have noticed that the variance of the loss increases (ie it starts to jump around in value), and it even begins to increase as we train longer. 

This phenomena is shown below. Y-axis is the loss, and the x-axis is the iteration number. At around interation ~30 I started getting the warning messages regarding "CG termination", and thus the curve starts to look more bumpy.

![six_output_learningcurve](https://user-images.githubusercontent.com/50096531/80552869-f9a9ab00-8995-11ea-9277-1bee94a91913.png)

My questions are:
- In layman's terms, what does the warning message "CG terminated in 1000 iterations" mean? I looked into the Conjugate Gradient method, and besides getting hit by a bunch of math, I got the sense that 1000 iterations isn't enough for this method, and some metric didn't hit a certain tolerance. Whatever it is, it definitely seems to negatively affect the training performance. 
- Is it possible that the increase in loss towards the end of the training is because of this "CG termination" warning? They definitely seem to be correlated, because the warnings starting popping up when the loss curve becomes "less smooth". If so, why? 
- The *bonus* question: what could be the cause of all of this? Is this a problem with my data? Or is the model just not meant for this kind of problem?

**Comment by jacobrgardner** [comment]
- The warning means the numerical methods we are using to compute the loss returned a potentially unacceptably inaccurate estimate of the loss.
- Yes, this could be causing the instability towards the end.
- Looking just at the magnitude of your loss values, this is almost certainly being caused by a lack of input data normalization. Try z-scoring your features and labels. That will probably fix the issue.

**Comment by christae** [comment]
Thanks for your quick reply. I had already normalized the labels and features, so I'm not sure if that was the issue, though I do wonder why the loss is so large. 

Regardless, I ended up changing the learning rate to 0.01 (before it was at 0.1), and changed the number of iterations to 1000. I still get the warning messages at around  iteration 750, however, the learning curve appears to be stable all throughout, and it shows consistent learning as the loss decreases. Here is the learning curve:

![image001](https://user-images.githubusercontent.com/50096531/80630316-04118680-8a22-11ea-9d66-ccd0f6cbdd72.png)

 Interestingly, I also noticed that the average residual norm as shown in the warning message now has a value around between 1 and 2. Before, the average residual norm would be a lot higher, in the tens, and even hundreds. 

Is there an explanation for why decreasing the learning rate to 0.01 somewhat solved the problem? The warning messages still show up, but it looks like the severity of the average residual norm has decreased.

**Comment by rachtsingh** [comment]
For future searchers, it's worth noting that one possible cause of ill-conditioned data (as mentioned here: https://github.com/cornellius-gp/gpytorch/issues/1297) is if you have highly similar data. In my case, that was because I had accidentally duplicated some data being passed in 😄.
